sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c

```diff
@@ -0,0 +1,541 @@
+        /* value buffer not big enough */
+        return -EOVERFLOW;
+    }
+
+    dev_eth_t *dev = netdev->ethdev;
+
+    dev->driver->set_promiscous(dev, (uint8_t)*value);
+
+    return sizeof(ng_netconf_enable_t);
+}
+
+static int _set(ng_netdev_t *dev, ng_netconf_opt_t opt, void *value,
+                size_t value_len)
+{
+    DEBUG("ng_netdev_eth: set ");
+
+    if ((dev == NULL) || (dev->driver != &ng_netdev_eth_driver)) {
+        DEBUG("[wrong device descriptor]\n");
+        return -ENODEV;
+    }
+
+    switch (opt) {
+        case NETCONF_OPT_PROMISCUOUSMODE:
+            DEBUG("promiscous mode\n");
+            return _set_promiscousmode((ng_netdev_eth_t *)dev, value, value_len);
+
+        default:
+            DEBUG("[not supported: %d]\n", opt);
+            return -ENOTSUP;
+    }
+}
+
+/* individual event handlers called by _isr_event() */
+static void _rx_event(ng_netdev_eth_t *dev);
+
+static void _isr_event(ng_netdev_t *dev, uint32_t event_type)
+{
+    DEBUG("ng_netdev_eth: ISR event ");
+
+    if ((dev == NULL) || (dev->driver != &ng_netdev_eth_driver)) {
+        return;
+    }
+
+    ng_netdev_eth_t *netdev = (ng_netdev_eth_t*)dev;
+
+    switch (event_type) {
+        case _ISR_EVENT_RX:
+
+            DEBUG("[ISR]\n");
+
+#ifdef NETDEV_ETH_DELAY_SEND
+            DEBUG("netdev_eth: delaying send...\n");
+            volatile int i = NETDEV_ETH_DELAY_SEND;
+            while(i--);
+            DEBUG("netdev_eth: delay done.\n");
+#endif
+
+            dev_eth_t *ethdev = netdev->ethdev;
+            ethdev->driver->isr(ethdev);
+            break;
+
+        default:
+            DEBUG("[unknown event_type]\n");
+            break;
+    }
+}
+
+static inline void _addr_set_broadcast(uint8_t *dst)
+{
+    memset(dst, 0xff, NG_ETHERNET_ADDR_LEN);
+}
+
+#define _IPV6_DST_OFFSET    (36)    /* sizeof(ipv6_hdr_t) - 4  */
+
+static inline void _addr_set_multicast(uint8_t *dst, ng_pktsnip_t *payload)
+{
+    switch (payload->type) {
+#ifdef MODULE_NG_IPV6
+        case NG_NETTYPE_IPV6:
+            dst[0] = 0x33;
+            dst[1] = 0x33;
+            memcpy(dst + 2, ((uint8_t *)payload->data) + _IPV6_DST_OFFSET, 4);
+            /* TODO change to proper types when ng_ipv6_hdr_t got merged */
+            break;
+#endif
+        default:
+            _addr_set_broadcast(dst);
+            break;
+    }
+}
+
+static int _marshall_ethernet(ng_netdev_eth_t *dev, uint8_t *buffer, ng_pktsnip_t *pkt)
+{
+    int data_len = 0;
+    ng_ethernet_hdr_t *hdr = (ng_ethernet_hdr_t *)buffer;
+    ng_netif_hdr_t *netif_hdr;
+    ng_pktsnip_t *payload;
+
+    if (pkt == NULL) {
+        DEBUG("ng_netdev_eth: pkt was NULL");
+        return -EINVAL;
+    }
+
+    payload = pkt->next;
+
+    if (pkt->type != NG_NETTYPE_NETIF) {
+        DEBUG("ng_netdev_eth: First header was not generic netif header\n");
+        return -EBADMSG;
+    }
+
+    hdr->type = byteorder_htons(ng_nettype_to_ethertype(pkt->next->type));
```

Comment:
Nice to have: add `NG_ETHERTYPE_RAW == 0xffff` if `hdr->type == 0`. Otherwise Wireshark shows garbage. Alternatively you can fix `ng_nettype_to_ethertype()` if this is desired in any case.